src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/AbstractRazorDelegatingEndpoint.cs

```diff
@@ -0,0 +1,113 @@
+using OmniSharp.Extensions.JsonRpc;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer
+{
+    internal abstract class AbstractRazorDelegatingEndpoint<TRequest, TResponse, TDelegatedParams> : IJsonRpcRequestHandler<TRequest, TResponse?>
+        where TRequest : TextDocumentPositionParams, IRequest<TResponse>
+        where TResponse : class?
+        where TDelegatedParams : class
+    {
+        private readonly DocumentContextFactory _documentContextFactory;
+        private readonly LanguageServerFeatureOptions _languageServerFeatureOptions;
+        private readonly RazorDocumentMappingService _documentMappingService;
+        private readonly ClientNotifierServiceBase _languageServer;
+        protected readonly ILogger _logger;
+
+        protected AbstractRazorDelegatingEndpoint(
+            DocumentContextFactory documentContextFactory,
+            LanguageServerFeatureOptions languageServerFeatureOptions,
+            RazorDocumentMappingService documentMappingService,
+            ClientNotifierServiceBase languageServer,
+            ILogger logger)
+        {
+            _documentContextFactory = documentContextFactory ?? throw new ArgumentNullException(nameof(documentContextFactory));
+            _languageServerFeatureOptions = languageServerFeatureOptions ?? throw new ArgumentNullException(nameof(languageServerFeatureOptions));
+            _documentMappingService = documentMappingService ?? throw new ArgumentNullException(nameof(documentMappingService));
+            _languageServer = languageServer ?? throw new ArgumentNullException(nameof(languageServer));
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+        }
+
+        /// <summary>
+        /// Override to provide behavior for razor contexts. Since the c#/html servers won't
+        /// have any information for this context it doesn't need to be delegated
+        /// </summary>
+        protected abstract Task<TResponse?> HandleInRazorAsync(TRequest request, CancellationToken cancellationToken);
+
+        /// <summary>
+        /// The delegated object to send to the <see cref="EndpointName"/>
+        /// </summary>
+        protected abstract Task<TDelegatedParams> CreateDelegatedParamsAsync(TRequest request, CancellationToken cancellationToken);
+
+        /// <summary>
+        /// The name of the endpoint to delegate to, from <see cref="LanguageServerConstants"/>
+        /// </summary>
+        protected abstract string EndpointName { get; }
+
+        /// <summary>
+        /// If the response needs to be remapped for any reasons, override to handle the remapping logic
+        /// </summary>
+        protected virtual Task<TResponse> RemapResponseAsync(TResponse delegatedResponse, DocumentContext documentContext, CancellationToken cancellationToken)
+            => Task.FromResult(delegatedResponse);
+
+        public async Task<TResponse?> Handle(TRequest request, CancellationToken cancellationToken)
```

Comment:
```suggestion
        public async Task<TResponse?> HandleAsync(TRequest request, CancellationToken cancellationToken)
```

Reply:
Sadly that name is part of the IRequestHandler interface from MediatR, which comes via the IJsonRpcRequestHandler interface from OmniSharp.

Reply:
Going to mark as resolved, thanks for answering David.